File: example/src/main/java/thirtytwo/degrees/halfpipe/example/cli/HelloCommand.java
package thirtytwo.degrees.halfpipe.example.cli;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.Options;
import org.apache.commons.lang.StringUtils;
import org.springframework.shell.core.CommandMarker;
import org.springframework.shell.core.annotation.*;
import org.springframework.stereotype.Component;
import thirtytwo.degrees.halfpipe.example.resources.Sayer;

import javax.inject.Inject;

/**
 * User: spencergibb
 * Date: 9/26/12
 * Time: 10:07 PM
 */
@Component
public class HelloCommand implements CommandMarker {

    @Inject
    Sayer sayer;

    @CliAvailabilityIndicator({"hello"})
    public boolean isCommandAvailable() {
        return true;
    }

    @CliCommand(value = "hello", help = "Say hello to the shell")
    public String hello(
            @CliOption(key = "message", mandatory = true, help = "The hello world message")
            String message,

            @CliOption(key = "more", mandatory = false,
                    help = "Have some more to say while saying hello",
                    specifiedDefaultValue="")
            String more,

            @CliOption(key = {"", "config"}, mandatory = false, help = "config file")
            String config )
    {
        StringBuilder hello = new StringBuilder(sayer.hello());
        hello.append(" with this message, "+message);
        if (!StringUtils.isBlank(more)) {
            hello.append(" "+more);
        }
        //System.out.println("\n\n\n\n"+hello+"\n\n\n\n");
        return hello.toString();
    }
}

File: halfpipe-scala-example/src/main/scala/thirtytwo/degrees/halfpipe/scalaexample/cli/HelloCommand.scala
package thirtytwo.degrees.halfpipe.scalaexample.cli

import javax.inject.{Named, Inject}
import com.netflix.config.scala.{DynamicStringProperty => DString}
import org.springframework.shell.core.CommandMarker
import org.apache.commons.lang.StringUtils
import org.springframework.shell.core.annotation.{CliOption, CliCommand, CliAvailabilityIndicator}
import org.springframework.stereotype.Component
import thirtytwo.degrees.halfpipe.scalaexample.ExampleScalaConfig

/**
 * User: spencergibb
 * Date: 10/2/12
 * Time: 9:53 PM
 */
@Component
class HelloCommand @Inject() (config: ExampleScalaConfig) extends CommandMarker {

  @CliAvailabilityIndicator(Array("hello"))
  def isCommandAvailable = true

  //TODO: write a scala dsl instead of shell annotations
  @CliCommand(value = Array("hello"), help = "Say hello to the shell")
  def hello(
             @CliOption(key = Array("message"), mandatory = true, help = "The hello world message")
             message: String,
             @CliOption(key = Array("more"), mandatory = false, help = "Have some more to say while saying hello",
             specifiedDefaultValue = "") more: String): String =
  {
    val hello: StringBuilder = new StringBuilder(config.helloText.get())
    hello.append(" with this message, " + message)
    if (!StringUtils.isBlank(more)) {
      hello.append(" " + more)
    }
    return hello.toString()
  }
}
